src/Symfony/Component/Cache/Adapter/MemcacheAdapter.php

```diff
@@ -0,0 +1,144 @@
+     * @param int         $defaultLifetime
+     */
+    public function __construct(\Memcache $client, $namespace = '', $defaultLifetime = 0)
+    {
+        parent::__construct($namespace, $defaultLifetime);
+        $this->client = $client;
+    }
+
+    /**
+     * Factory creation method that provides an instance of this adapter with a\Memcache client instantiated and setup.
+     *
+     * Valid DSN values include the following:
+     *  - memcached://localhost                  : Specifies only the host (defaults used for port and weight)
+     *  - memcached://example.com:1234           : Specifies host and port (defaults weight)
+     *  - memcached://example.com:1234?weight=50 : Specifies host, port, and weight (no defaults used)
+     *
+     * @param string|null $dsn
+     *
+     * @return MemcacheAdapter
+     */
+    public static function create($dsn = null)
+    {
+        if (!extension_loaded('memcache')) {
+            throw new InvalidArgumentException('Failed to create Memcache client due to missing "memcache" extension.');
+        }
+
+        $adapter = new static(new \Memcache());
+        $adapter->setup($dsn ? array($dsn) : array());
+
+        return $adapter;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doSave(array $values, $lifetime)
+    {
+        $result = true;
+
+        foreach ($values as $id => $val) {
+            $result = $this->client->set($id, $val, null, $lifetime) && $result;
+        }
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doFetch(array $ids)
+    {
+        foreach ($this->client->get($ids) as $id => $val) {
+            yield $id => $val;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doHave($id)
+    {
+        return $this->client->get($id) !== false;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doDelete(array $ids)
+    {
+        $remaining = array_filter($ids, function ($id) {
+            return false !== $this->client->get($id) && false === $this->client->delete($id);
+        });
+
+        return 0 === count($remaining);
+    }
+
+    protected function getAllIdsMatching($namespace)
+    {
+        $ids = array();
+        foreach ($this->client->getExtendedStats('slabs') as $slabGroup) {
+            foreach ($slabGroup as $slabId => $slabMetadata) {
+                if (!is_array($slabMetadata)) {
+                    continue;
+                }
+                foreach ($this->client->getExtendedStats('cachedump', (int) $slabId, 0) as $slabIds) {
+                    if (is_array($slabIds)) {
+                        $ids = array_merge($ids, array_keys($slabIds));
+                    }
+                }
+            }
+        }
+
+        return array_filter((array) $ids, function ($id) use ($namespace) {
+            return 0 === strpos($id, $namespace);
+        });
+    }
+
+    protected function addServer($dsn)
```

Comment:
all protected method should be private in fact (we are usd to think twice before accepting any public or protected method because they are covered by our BC promise)